Source/Initialization/WarpXInitData.cpp

```diff
@@ -695,46 +695,172 @@ WarpX::InitLevelData (int lev, Real /*time*/)
+            // we need 1 more ghost cell than Mfield_fp has because
+            // we are averaging to faces, including the ghost faces
+            BoxArray bx = Mfield_fp[lev][0]->boxArray();
+            amrex::MultiFab Mx(bx.enclosedCells(), Mfield_fp[lev][0]->DistributionMap(), 1, Mfield_fp[lev][0]->nGrow()+1);
+            amrex::MultiFab My(bx.enclosedCells(), Mfield_fp[lev][1]->DistributionMap(), 1, Mfield_fp[lev][1]->nGrow()+1);
+            amrex::MultiFab Mz(bx.enclosedCells(), Mfield_fp[lev][2]->DistributionMap(), 1, Mfield_fp[lev][2]->nGrow()+1);
+
+            // Initialize Mfield_fp with external function
+            InitializeExternalFieldsOnGridUsingParser(&Mx,
+                                                      &My,
+                                                      &Mz,
+                                                      getParser(Mxfield_parser),
+                                                      getParser(Myfield_parser),
+                                                      getParser(Mzfield_parser),
+                                                      lev);
+
+            // average Mx, My, Mz to faces in Mfield_fp
+            for (MFIter mfi(*Mfield_fp[lev][0], TilingIfNotGPU()); mfi.isValid(); ++mfi) {
+
+                const amrex::Box& tbx = mfi.tilebox( IntVect(1,0,0), Mfield_fp[lev][0]->nGrowVect() );
+                const amrex::Box& tby = mfi.tilebox( IntVect(0,1,0), Mfield_fp[lev][1]->nGrowVect() );
+                const amrex::Box& tbz = mfi.tilebox( IntVect(0,0,1), Mfield_fp[lev][2]->nGrowVect() );
+
+                auto const& mx_cc = Mx.array(mfi);
+                auto const& my_cc = My.array(mfi);
+                auto const& mz_cc = Mz.array(mfi);
+
+                auto const& m_xface = Mfield_fp[lev][0]->array(mfi);
+                auto const& m_yface = Mfield_fp[lev][1]->array(mfi);
+                auto const& m_zface = Mfield_fp[lev][2]->array(mfi);
+
+                amrex::ParallelFor (tbx, tby, tbz,
+                [=] AMREX_GPU_DEVICE (int i, int j, int k) {
+                    m_xface(i,j,k,0) = 0.5*(mx_cc(i-1,j,k) + mx_cc(i,j,k));
+                    m_xface(i,j,k,1) = 0.5*(my_cc(i-1,j,k) + my_cc(i,j,k));
+                    m_xface(i,j,k,2) = 0.5*(mz_cc(i-1,j,k) + mz_cc(i,j,k));
+                },
+                [=] AMREX_GPU_DEVICE (int i, int j, int k) {
+                    m_yface(i,j,k,0) = 0.5*(mx_cc(i,j-1,k) + mx_cc(i,j,k));
+                    m_yface(i,j,k,1) = 0.5*(my_cc(i,j-1,k) + my_cc(i,j,k));
+                    m_yface(i,j,k,2) = 0.5*(mz_cc(i,j-1,k) + mz_cc(i,j,k));
+                },
+                [=] AMREX_GPU_DEVICE (int i, int j, int k) {
+                    m_zface(i,j,k,0) = 0.5*(mx_cc(i,j,k-1) + mx_cc(i,j,k));
+                    m_zface(i,j,k,1) = 0.5*(my_cc(i,j,k-1) + my_cc(i,j,k));
+                    m_zface(i,j,k,2) = 0.5*(mz_cc(i,j,k-1) + mz_cc(i,j,k));
+                });
```

Comment:
May be after this PR is merged, we can see if we can use the Interp function directly and use this method as a way to compare the results?. If it is okay, I will add that the Interp function for init as an issue? That way, we dont have a built-in assumption about Mfield staggering.

Reply:
Yes, for future PR.  I just wanted to get something working in there.

Reply:
Yes, for future PR. I just wanted to get something working in there.